cfpb/cfgov-refresh
# Add resource snippets and snippet list module

This PR defines the Resource snippet type and creates a Snippet List module to be able to output them in a filtered, compact format.

**We'd very much like to get this in tomorrow's 4.6 code freeze.** Sorry for the short turnaround, but the code has already been fairly thoroughly vetted via pairing with @chosak and @jimmynotjim and the concept by @schaferjh.

## Additions

- Resources snippet definition
- Snippet List module

## Changes

- Fixes Image and Text 25/75 output to not output an empty `h3` if `heading` is not set.
- Simplifies (and also makes more complex) the `.o-table__stack-on-small` code

## Testing

1. Pull branch and check it out
2. `./cfgov/manage.py migrate`
3. `./cfgov/manage.py runserver`
4. Go to the Snippets screen and create a new Resource, making sure to add a related file or link, and giving it a couple tags. **Remember the tag names you gave it.**
5. Open or create a browse page
6. Create a Snippet List module in the main content area. Be sure to select "Resources" for the snippet type and enter tag names to match the tags you gave the snippet you created. Also, create at least one action, choosing the snippet field that corresponds to to the related file or link field you entered on your snippet.
7. Preview the page

## Screenshots

![screen shot 2017-01-18 at 13 44 19](https://cloud.githubusercontent.com/assets/1044670/22078098/e4a7b316-dd84-11e6-8c83-d21a0cbb395c.png)

![screen shot 2017-01-18 at 13 45 30](https://cloud.githubusercontent.com/assets/1044670/22078103/e6ae9f1c-dd84-11e6-9492-291ca414679f.png)

![screen shot 2017-01-18 at 13 46 39](https://cloud.githubusercontent.com/assets/1044670/22078108/eba75c34-dd84-11e6-94c4-886024ee61e7.png)

## Notes

- The tag-based filtering mechanism isn't ideal, but it's a working MVP. Ideally, we'd have a hard-coded list of snippet (sub)categories for editors to choose from, but until we get closer to continuous delivery and/or our idea of what tags gets more solidified, having to deploy code to update that list would be extremely painful.

## Todos

- [x] Write tests for the new `filter_by_tags` method on `TaggableSnippetManager`.

## Checklist
